### ISSUE TITLE
Documentation Updates/Expansion Ideas

### ISSUE THREAD

**Issue by dlar1990** [Bionus/imgbrd-grabber#2030]
Full disclosure:  I'm only just smart enough to be dangerous, and I'm running on very little sleep right now, so pardon me if I ramble or say anything stupid/obvious/incorrect here ;)

Current documentation is, somewhat understandably,, a bit lacking in terms of covering more advvanced features, as well as how Grabber actually does its work.  I find myself in the predicament of having Grabbed a large number of files, with little information readily available on how to work with them now that I have them squirreled away due to some key information being missing, tucked away in strange places, and/or overly specific.

For instance;

1. Given the current state of options available for hosting boorus (poorly optimized, no longer actively maintained, highly involved in setup, lacking features... etc) I've considered the possibility of using my frustration as an excuse to extend my Python knowledge and try to write up a more easily deploy-able local network oriented solution.  The only problem is, I'm not too sure on how the data set I'd be referencing as a model (namely tags and other externally stored "metadata") is being stored and utilized by Grabber.

2. Documentation (as well as discussions in the Issues section) seem largely predicated on handling things _before_ accumulating a stash of files.  If someone chooses to migrate solutions for working with their data, things get dicey/obfuscated.  

3. The Booru setup docs seem to assume a lot about the context in which one may wish to utilize them.   For instance, they seem largely tailored to use specific tools like WAMP or a VM and detail how to set up post-download jobs to integrate newly downloaded files into the appropriate DBs.  This conflicts with my use case in that I have a lot of images I've already downloaded, and an existing Apache instance set up via [FreedomBox]([url]http://freedombox.org/(url)) which I'd love to 'strap onto for hosting a Booru as a home for all of this lovely data.  Unfortunately my attempts at adapting the current documentation have proven fruitless.

4. It seems Hydrus is a common tool used by many Grabber users, but all of the information on how to get the two playing nicely together seems to be scattered across the Issues section here.  Formalized documentation on this would likely be very useful, especially if someone familiar with the workings of both the standalone client and server use cases would be willing & able to share their knowledge.

5. There's not a super clear path from "how to download files" to "how to view files you've downloaded."  The Booru docs are sort of floating in the "commands" category (???) without a solid correlative link in other sections, and post-filtering for metatags like grabber:downloaded is tucked away under the + with only a brief mention in the "Getting Started" section.  A clarifying line in the "Search" section itself about that would likely be helpful, as would creating a "hosting your Grabbed images" or similarly named section.  Side note: I do wonder if it would be useful from an end-user experience perspective to add something like a checkbox, a secondary type of search tab, etc to handle such filtering.

Side query:  Does the built-in image viewer still query the original source for an image if it's already on disk?  I would expect not, but I do find that occasionally it takes a comparable amount of time to render an image I've already downloaded as it would for an API request to fetch it if I hadn't.

TLDR: More of a peak under the hood without having to dig into the source code, tracking down files manually, etc would be invaluable, as would some slight tweaks to make documentation of advanced features and more general use cases more intuitive.

P.S. - If I ever do get the project mentioned in Pt #1 off of the ground, I fully intend to explicitly build it to primarily interface with Grabber's model for tag storage etc, obviously with appropriate attribution etc.  Literally every other tool is hilariously wonky or limited compared to what you've build here, and anything I'd care to build will be focused on bringing a similar ease of use for normal end users to the booru hosting side of things. My sincere gratitude to all of you for your hard work! <3

**Comment by Bionus** [comment]
> Given the current state of options available for hosting boorus (poorly optimized, no longer actively maintained, highly involved in setup, lacking features... etc) I've considered the possibility of using my frustration as an excuse to extend my Python knowledge and try to write up a more easily deploy-able local network oriented solution.

You might be interested in issue #1654, but recently Szurubooru was added, which is trivial to install and have working with Grabber:
https://bionus.github.io/imgbrd-grabber/docs/commands/szurubooru.html

> The only problem is, I'm not too sure on how the data set I'd be referencing as a model (namely tags and other externally stored "metadata") is being stored and utilized by Grabber.

Not sure what you mean. Grabber can be set up to send any kind of "command" every time an image is downloaded, either an SQL or a process command, and either at the image or at the tag level. Whatever model you choose to work with, it should be easy to have Grabber write into them. Especially since there's aren't 3,000 ways to design storage of images with tags in a database.

> Documentation (as well as discussions in the Issues section) seem largely predicated on handling things before accumulating a stash of files. If someone chooses to migrate solutions for working with their data, things get dicey/obfuscated.

Because that's what Grabber does: it downloads images. It provides a few side things for integrating with other tools more easily, but it's not designed to manage your collection. Many tools already exist for that.

> The Booru setup docs seem to assume a lot about the context in which one may wish to utilize them. For instance, they seem largely tailored to use specific tools like WAMP or a VM and detail how to set up post-download jobs to integrate newly downloaded files into the appropriate DBs. This conflicts with my use case in that I have a lot of images I've already downloaded, and an existing Apache instance set up via FreedomBox which I'd love to 'strap onto for hosting a Booru as a home for all of this lovely data. Unfortunately my attempts at adapting the current documentation have proven fruitless.

Setting up the booru is the user's responsibility. I only added the steps in the docs for a few popular ones (notably Gelbooru and Shimmie) because their setup had a few quircks that some users might not know. If you want to install a booru on a real server, on a Raspberry PI, or wherever else, it's expected of you to make the appropriate changes. Most users only want to manage their local collection, so the basic tutorials are doing exactly that. It's expected that if you can setup a server, you should be able to install a basic booru like Gelbooru or Szurubooru on one as well. If you can't / face issues, you can always ask for help on the relevant booru's forum / issues.

> It seems Hydrus is a common tool used by many Grabber users, but all of the information on how to get the two playing nicely together seems to be scattered across the Issues section here. Formalized documentation on this would likely be very useful, especially if someone familiar with the workings of both the standalone client and server use cases would be willing & able to share their knowledge.

I don't use Hydrus personally, hence the lack of documentation, but if someone knows, PRs to add basic information about it are always welcome.

> There's not a super clear path from "how to download files" to "how to view files you've downloaded."

As said above, that's because for most users, you'll see the files directly in your file explorer, and that'd be enough. Only "power-users" will use Hydrus or a local booru to manage their downloaded files. And while those work well with Grabber, they're not part of it.

> The Booru docs are sort of floating in the "commands" category (???) without a solid correlative link in other sections, [...] creating a "hosting your Grabbed images" or similarly named section. 

I guess the introduction / titles could be improved for this part.

> and post-filtering for metatags like grabber:downloaded is tucked away under the + with only a brief mention in the "Getting Started" section. A clarifying line in the "Search" section itself about that would likely be helpful

:+1:

> Side note: I do wonder if it would be useful from an end-user experience perspective to add something like a checkbox, a secondary type of search tab, etc to handle such filtering.

I feel that would just be more confusing.
